File: Python/Problems/Problems/problem2946.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem2946(Problem):
    def test(self) -> bool:
        mat = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
        k = 4

        expected = False

        result = self.areSimilar(mat, k)

        print(result)

        return result == expected

    def areSimilar(self, mat: List[List[int]], k: int) -> bool:
        n = len(mat)
        m = len(mat[0])
        if k == m or k == 0:
            return True
        for i in range(n):
            for j in range(m):
                if i % 2 == 0:
                    if mat[i][j] != mat[i][((j - k) % m + m) % m]:
                        return False
                else:
                    if mat[i][j] != mat[i][(j + k) % m]:
                        return False

        return True

File: C++/Problems/Problems/problem_2946.cpp
#include "problem_2946.h"

bool problem_2946::test() {
    std::vector<std::vector<int>> mat {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int k = 4;

    bool expected = false;

    auto result = areSimilar(mat, k);

    std::cout << result << std::endl;

    return result == expected;
}

bool problem_2946::areSimilar(std::vector<std::vector<int>> &mat, int k) {
    int n = mat.size();
    int m = mat[0].size();
    if (k == m || k == 0)
        return true;
    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < m; ++j)
        {
            if (i % 2 == 0)
            {
                if (mat[i][j] != mat[i][((j - k) % m + m) % m])
                    return false;
            }
            else
            {
                if (mat[i][j] != mat[i][(j + k) % m])
                    return false;
            }
        }
    }

    return true;
}
